### PR TITLE
Openzfs compressedarc abd patchset WIP

### PR DESCRIPTION
**Pull request by dpquigl**
This is a WIP PR for the abd work. This follows the path outlined by Dan Kimmel at delphix and contains multiple features including compressed arc, compressed send/recv, a reworking of arc internals, and the ABD code with some linux specific patches.

### CODE REVIEW

**Comment by ironMann** [comment]
@dpquigl I've been rebasing SIMDized raidz patches on couple of other ABD patch sets. Last attempt was branch [raidz-abd](https://github.com/ironMann/zfs/commits/raidz-abd)

However, that requires a slight change in ABD patch, most notably making `abd_get_offset()` accept `size` parameter. Commit: 84adc455ca2f674888125b05d19ec310d1d5a279 Is there a way to push this change upstream?

**Comment by dpquigl** [comment]
@ironMann It looks like your work is against David Chen's version of ABD. This version we have here is against the Delphix-OS version worked on by Dan Kimmel. I'll look to see what its trying to return with that size parameter but its possible it doesn't make sense with the other ABD implementation.

**Comment by behlendorf** [comment]
@ironMann that slight change to abd_get_offset() looks entirely reasonable to me.  I can definitely see how it would make adding ABD to the vectorized raidz simpler.  Let's get @dankimmel's thought on this too make sure it's not going to cause problems elsewhere.

**Comment by ironMann** [comment]
@dpquigl To clarify the size parameter: When a new _overlay abd_ is created with `abd_get_offset()` it inherits buffer end from the host, meaning its size is `(host_end - offset)`. For raidz, it is beneficial to split up one io abd into several consecutive abds. By providing offset and size, those can behave as regular abds under iterator methods.

**Comment by dankimmel** [comment]
@ironMann: neat, this optimization looks really cool!

@dpquigl is right, the new ABD code is a significant departure from the version you patched. The iterator functions have all been rewritten in the new version to use `abd_iterate_func()` and `abd_iterate_func2()` under the hood, but obviously your version is implementing some new primitives in `abd.c` to handle iterating over groups of >2 ABDs. Is is plausible that your `abd_raidz_gen_iterate()` and `abd_raidz_rec_iterate()` functions could be rewritten using those, or maybe rewrite `abd_iterate_func2()` to take a variable number of ABDs? I don't know enough about the `func_raidz_gen`/`func_raidz_rec` implementations you're passing into those to know for sure what the best path might be.

Why do you need the size stored for offset ABDs -- just because it would be easier to track the lengths of the segments to compare? It shouldn't be hard to add that, just trying to understand the use case better, since I think you _could_ achieve the same thing by passing the length to `abd_iterate_func*` instead.

**Comment by ironMann** [comment]
@dankimmel @behlendorf I've ported raidz onto this abd stack, PR #5020.

2d660f3109d0cfe4005f003b7f58a727867dbcad is adaptation of ABD needed for the rest of the RAIDZ code. There are few minor changes, and two special iterator methods.

> Is it plausible that your abd_raidz_gen_iterate() and abd_raidz_rec_iterate() functions could be rewritten...

I'm already using `abd_iterate_func()` and `abd_iterate_func2()`, but without access to more buffers at once performance would suffer. Ideally, I would simultaneously access as many abd buffers as there are devices in raidz map. I know that this breaks layering, but teh methods are somewhat special, and if need be, I can reimplement them elsewhere. Problem is, ABD iterator interface is not exported.

> Why do you need the size stored for offset ABDs

For iterators (created from offset ABDs) in `abd_raidz_gen_iterate()` and `abd_raidz_rec_iterate()` to work correctly :)
Please also review this [change](https://github.com/zfsonlinux/zfs/pull/5020/commits/2d660f3109d0cfe4005f003b7f58a727867dbcad#diff-c0e54e4510dcf312249c8bf80e635187R786) for `iter_mapsize` calculation of scatter buffers.

File: module/zfs/abd.c [hunk 1]
```diff
@@ -215,15 +221,37 @@ abd_fini(void)
...
+
+#else
 
-	kmem_cache_destroy(abd_chunk_cache);
-	abd_chunk_cache = NULL;
+struct page;
+#define	kpm_enable			1
+#define	abd_alloc_chunk() \
+	((struct page *)kmem_alloc(PAGESIZE, KM_SLEEP))
+#define	abd_free_chunk(chunk) 		kmem_free(chunk, PAGESIZE)
```
Comment (behlendorf) [inline module/zfs/abd.c]: Prior to this change PAGESIZE wasn't used extensively in userspace so it was implemented in `lib/libspl/include/sys/param.h` just as `sysconf(_SC_PAGESIZE)`.  That's not going to work well if use PAGESIZE everywhere.  We're either going to want to implement PAGESIZE differently or alternately preserve the  `zfs_abd_chunk_size` and initialize it once in `abd_init()`.

Potentially we could also preserve the kmem cache chunk case for use solely by user space.  That would help minimize our delta with upstream.

**Comment by dankimmel** [comment]
> Ideally, I would simultaneously access as many abd buffers as there are devices in raidz map
Yeah, I agree that this is such a special case that adding them to `abd.c` is fine. If we accrue more callers like this in the future then we can think harder about whether there's a way to abstract the underlying ABD iteration, or whether we should just expose that externally like you suggested.
> For iterators (created from offset ABDs) in `abd_raidz_gen_iterate()` and `abd_raidz_rec_iterate()` to work correctly :)
Heh. I see what you're saying now.
> Please also review this change for `iter_mapsize` calculation of scatter buffers.
The `iter_mapsize` calculation looks right to me, and adding the size with the `_impl` function looks good too.

File: module/zfs/abd.c (hunk 1 above)
Comment (ironMann) [inline module/zfs/abd.c]: @behlendorf Good point. How reliant is userspace to the actual running system's `PAGE_SIZE`? Can we get it in configure script, and define `PAGE_SIZE` then?

I'm now more worried about kernel's `PAGESIZE` value using the build system's kernel [value](https://github.com/zfsonlinux/spl/blob/master/include/sys/sysmacros.h#L111). `PAGE_SHIFT` is configurable on some arch, e.g. arm64 and ia64.

**Comment by behlendorf** [comment]
@dpquigl the patch in #5024 resolves the Ubuntu and Debian buildbot failures.  You'll want to pull it in to this branch at least until it's merged to master.

File: include/sys/spa.h [hunk 2]
```diff
@@ -577,7 +577,6 @@ extern int spa_get_stats(const char *pool, nvlist_t **config, char *altroot,
     size_t buflen);
 extern int spa_create(const char *pool, nvlist_t *config, nvlist_t *props,
     nvlist_t *zplprops);
-extern int spa_import_rootpool(char *devpath, char *devid);
```
Comment (behlendorf) [inline include/sys/spa.h]: Good find, we accidentally missed this declaration.

File: module/zfs/abd.c (hunk 1 above)
Comment (behlendorf) [inline module/zfs/abd.c]: Getting `PAGE_SIZE` wrong in user space shouldn't be terrible.  I could see it perhaps causing some memory alignment issues but I suspect that would be the worst of it.  That said I've never tried.

On the kernel side we should always be using `PAGE_SIZE` from the kernel headers.  I don't recall why we ever added the `#ifdef` around PAGESIZE, it seems like we'd want it to fail if `PAGESIZE` was ever defined.

Comment (ironMann) [inline module/zfs/abd.c]: Potential `PAGE_SHIFT` discrepancies will be resolved by dkms. I was thrown off by a colleague's driver debugging story with two arm64 systems with different page sizes.

**Comment by behlendorf** [comment]
Closing.  To be replaced by a new rebased PR with this changes now that compressed ARC and compressed send/recv have been merged.

**Comment by mmatuska** [comment]
MRU/MFU pressure fix:
#10548
#10618

**Pull request closed without merging**